### PR TITLE
Convert `validate.sh` to a Haskell script

### PR DESCRIPTION
**Pull request by 9999years**
Closes https://github.com/haskell/cabal/issues/10317.

A Haskell script will be easier to maintain and expand than the existing Bash script.

This also adds a `--pattern PATTERN` option which lets you filter tests across all test suites.

I've kept the new script's behavior as close to the old script's behavior as possible, to reduce design discussion on this PR.

### TODO

- [x] Add `cabal-validate/README.md` to make it obvious where to look for logic and how to interact with the program
- [x] Add Haddock comments to most definitions

### Future work

- It should be possible to override the default `tasty` args (currently `--hide-successes`)
- `--list-steps` only lists the steps that are enabled with the given command-line options, not all possible steps; `./validate.sh --list-steps` and `./validate.sh --list-steps --run-cli-suite` give different results! (This is an existing bug in the old script.)
- Perhaps "steps" should be divided into "suites"; for example, the `lib-tests` step runs five different test suites, making it take much longer than necessary to reproduce test failures.
- Commands printed to users should be shell escaped. (Exiting bug in the old script.)
- `--help` output could use some love; it wasn't very clear to me what various commands do.
- `--keep-going` mode to ignore failures and continue executing tests/suites.
- The output is _really_ verbose, because the command lines are so long.

### CODE REVIEW

**Comment by ulysses4ever** [comment]
I wonder if it can be just a [cabal script](https://cabal.readthedocs.io/en/3.12/cabal-commands.html#cabal-run)... But this is not crucial at the moment.

**Comment by philderbeast** [comment]
I hope I'm not jumping the gun with a pull request in draft but HLint is gagged[^1] and could help.

[^1]:  The gag order is #9110. We want to protect existing code from churn but this is new code.

We suppress a lot of HLint suggestions at the root :-(

https://github.com/haskell/cabal/blob/3a6f73ed7286ff69702bb83233a7b328f39c966b/.hlint.yaml#L1-L95

Side-stepping that configuration that ignores 94 suggestions lets HLint come forward with a few suggestions:

```
$ sed 's/^- ignore/# &/' .hlint.yaml > .hlint-open.yaml \
  && hlint cabal-validate/main/Main.hs --hint=.hlint-open.yaml
cabal-validate/main/Main.hs:125:26-52: Suggestion: Use :
Found:
  ["build"] ++ cabalArgs opts
Perhaps:
  "build" : cabalArgs opts

cabal-validate/main/Main.hs:128:25-54: Suggestion: Use :
Found:
  ["list-bin"] ++ cabalArgs opts
Perhaps:
  "list-bin" : cabalArgs opts

cabal-validate/main/Main.hs:213:12-60: Warning: Use maybe
Found:
  fromMaybe getNumCapabilities (pure <$> opts . jobs)
Perhaps:
  maybe getNumCapabilities pure (opts . jobs)

cabal-validate/main/Main.hs:303:12-43: Suggestion: Redundant bracket
Found:
  (option (maybeReader parseStep))
    (long "step"
       <>
         help "Run only a specific step (can be specified multiple times)")
Perhaps:
  option
    (maybeReader parseStep)
    (long "step"
       <>
         help "Run only a specific step (can be specified multiple times)")

cabal-validate/main/Main.hs:(373,1)-(374,66): Warning: Eta reduce
Found:
  boolOption defaultValue trueName modifiers
    = boolOption' defaultValue trueName ("no-" <> trueName) modifiers
Perhaps:
  boolOption defaultValue trueName
    = boolOption' defaultValue trueName ("no-" <> trueName)

cabal-validate/main/Main.hs:472:16-42: Suggestion: Use tuple-section
Found:
  \ exitCode -> (exitCode, "")
Perhaps:
  (, "")
Note: may require `{-# LANGUAGE TupleSections #-}` adding to the top of the file

6 hints
```

**Comment by philderbeast** [comment]
> I wonder if it can be just a [cabal script](https://cabal.readthedocs.io/en/3.12/cabal-commands.html#cabal-run)... But this is not crucial at the moment.

@ulysses4ever, it could be but unfortunately it can't also be a script. It can't be both an installable executable and a script (but I wish it could), reported as #10324.

File: cabal-validate/main/Main.hs
Comment (geekosaur) [inline cabal-validate/main/Main.hs:368]: In general, this distinguishes `Cabal` (with `Cabal-syntax` being considered a sub-component) from `cabal-install` (aka "cli"). This also applies to the later ones.

File: cabal-validate/main/Main.hs
Comment (geekosaur) [inline cabal-validate/main/Main.hs:348]: It's the `cabal` being tested; the `cabal` on `$PATH` is used for builds.

File: cabal-validate/main/Main.hs
Comment (geekosaur) [inline cabal-validate/main/Main.hs:342]: For builds; `extra-hc` below is used for tests.

File: cabal-validate/main/Main.hs
Comment (geekosaur) [inline cabal-validate/main/Main.hs:738]: I think you're looking for [`cabal-plan`](https://hackage.haskell.org/package/cabal-plan). At some point this should be made an external command (will require some option processing changes, see https://github.com/haskell/cabal/issues/10275). I don't know if it's worth integrating into `cabal-install` itself, especially given the dependency footprint of `aeson`.

**Review by geekosaur — APPROVED** [review]

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:58]: Why not https://hackage.haskell.org/package/text-2.1.1/docs/Data-Text-IO.html#v:putStrLn?

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:114]: Could it be
```suggestion
          | (parsed, []) <- parsedVersions
```
?

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:106]: Is it `turtle` forcing us to convert between `Text` and `String` everywhere? Is there enough added value in comparison to `process`?

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:326]: Does it switch between `False` and `False`?

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:387]: Same question.

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:517]: Could `terminal-size` to be imported qualified? Otherwise it's hard to guess what kind of `size` it is.

File: cabal-validate/main/Main.hs
Comment (Bodigrim) [inline cabal-validate/main/Main.hs:525]: What would happen if I redirect output to a file? Will it be all riddled with ANSI codes? Could we check `hSupportsANSI`?

File: cabal-validate/main/Main.hs
Comment (9999years) [reply cabal-validate/main/Main.hs:326]: Yes, this preserves the behavior of `validate.sh`. I agree that this is a bit silly, but I didn't want to change anything in this PR.

File: cabal-validate/main/Main.hs
Comment (9999years) [reply cabal-validate/main/Main.hs:525]: > Will it be all riddled with ANSI codes?

Yes, this is what the existing `validate.sh` script does as well. The behavior hasn't changed since it was added in 2018 ([`6a9a101`](https://github.com/haskell/cabal/commit/6a9a101ca5e1731b9f099a94579a84d2e19667f8#diff-fe5f603ef22a6b8b43d317780c6c5a4bd83d130705a9abfae8cf4e968ce4dbd2R54)).

> Could we check `hSupportsANSI`?

That's a good idea, could you file an issue for that? I'd like to keep behavior changes out of this PR.

**Comment by ulysses4ever** [comment]
> @[9999years](https://github.com/9999years) requested review from [philderbeast](https://github.com/philderbeast) and [ulysses4ever](https://github.com/ulysses4ever)

It's on my radar but I'll only get to it the next week I'm afraid

File: validate.sh
Comment (geekosaur) [inline validate.sh:3]: Do we need to force `bash` for a one-liner? The only change needed is to replace `"$@"` with `${1+"$@"}`.

**Review by geekosaur — APPROVED** [review]

File: cabal-validate/main/Main.hs
Comment (9999years) [reply cabal-validate/main/Main.hs]: Done, switched over to `typed-process` and the code is a lot nicer now. It's a shame that `text` isn't the standard though, UTF-8 is really so much better than a linked list of 4-byte code units.

**Comment by 9999years** [comment]
OK, I've readied the PR for review and split it into modules. This will keep the script's logic (aside from CLI parsing) in the `Main.hs` module and the boilerplate separate.

**Review by geekosaur — APPROVED** [review]

**Comment by 9999years** [comment]
Rebased to fix conflicts, no changes.

**Comment by geekosaur** [comment]
It'll be a bit before I get to re-reviewing, I have other things to do today. (Does it really need to be reapproved?)

**Comment by 9999years** [comment]
OK, I've also added Haddock comments to all definitions to make the script friendly for newcomers and easy to maintain, and I wrote a short `README.md` to point browsers to the `validate.sh` entrypoint or the `Main.hs` entrypoint depending on their needs.

**Review by geekosaur — APPROVED** [review]

**Comment by 9999years** [comment]
Hi, still waiting for a second review on this (and my other PRs).

**Comment by ulysses4ever** [comment]
I'll try to look at it on the weekend. Sorry it's taking so long...

File: cabal-validate/src/Main.hs
Comment (ulysses4ever) [inline cabal-validate/src/Main.hs:207]: The doctest step in validate.sh (which is what you were inspired by, I guess) has bit rotted. Look at how we call doctest in the root Makefile. That’s what doctest CI job uses.  Perhaps, that logic should move to validate and the Makefile should only call the corresponding validate step.

File: cabal-validate/src/Main.hs
Comment (ulysses4ever) [inline cabal-validate/src/Main.hs:283]: I really hope we can find a way to externalise this value at some point. Just a text file with the timestamp and a descriptive name, which cabal-validate could read in, perhaps?

File: cabal-validate/src/Main.hs
Comment (ulysses4ever) [inline cabal-validate/src/Main.hs:153]: This is purely stylistic and subjective, but I’d much rather not read `\n`. Luckily, `unlines` does the job. (As a bonus, half of `<>` will turn into `,` — what a profit we get if we were paid by how much we’re saving on characters!..)

File: cabal-validate/src/Main.hs
Comment (ulysses4ever) [inline cabal-validate/src/Main.hs:127]: Another stylistic wibble that you don’t have to address: I dislike the plentiful applications to `opts` everywhere. My suggestion would be to use NamedFieldPuns and destruction of `opts` in the argument position.

File: cabal-validate/src/Main.hs
Comment (geekosaur) [reply cabal-validate/src/Main.hs:283]: If we didn't have to deal with 3.10.3's broken `import`, that would be easy to do. As is, it's nightmarish to get right in a way that works with all cabal versions in use for Cabal development.

File: cabal-validate/src/Main.hs
Comment (ulysses4ever) [inline cabal-validate/src/Main.hs:364]: Could replace the ++ with multiple :-applications. I think we have some hlint integration but it probably has a lot of stuff turned off because we didn’t want to update the whole codebase at once. May be good to research how to make new packages like cabal-validate get covered by hlint without (with minimal) exceptions.

File: cabal-validate/src/Step.hs
Comment (ulysses4ever) [inline cabal-validate/src/Step.hs:27]: Do we really need Show? Given that there’s no Read and that there’s a dedicated display function.

Comment (geekosaur) [reply cabal-validate/src/Step.hs:27]: Ability to debug is nice.

File: cabal-validate/src/Cli.hs
Comment (ulysses4ever) [inline cabal-validate/src/Cli.hs:62]: I’d prefer an empty line after a haddock and the following field. Otherwise, especially when haddock has more than one line, it’s a little hard to associate it right during skimming.

File: cabal-validate/src/Cli.hs
Comment (ulysses4ever) [inline cabal-validate/src/Cli.hs:294]: I’d expect this be named `rawOptsParser` (likewise for the “full” version below).

File: cabal-validate/src/Cli.hs
Comment (ulysses4ever) [inline cabal-validate/src/Cli.hs:270]: I’d re-iterate the difference with `Opts`, e.g.:
```suggestion
-- | Literate command-line options as supplied by the user, before resolving the environment.
```

File: cabal-validate/src/OutputUtil.hs
Comment (ulysses4ever) [inline cabal-validate/src/OutputUtil.hs:59]: I was surprised that this guy receives the whole kitchen sink of Opts. If it only uses the initial start time, it may be better to pass exactly that. Otherwise, it reminds me the dreaded DynFlags in GHC and how much pain it took to tear them apart (see the GHC modularity document from a couple of years ago).

**Review by ulysses4ever — APPROVED** [review]
Amazing work, thanks!

My only big concern is how many times it will have to be built from scratch and whether it will bring a noticeable slowdown in CI. The dependency footprint looks non-negligible and hence I expect noticeable build times. Perhaps, it’d be wiser to have it as a standalone GitHub project that could host its own binary releases, which we could download during cabal CI. Or something fancy like nix-based build and cachix.

**Comment by ulysses4ever** [comment]
I removed the merge-related labels for the time being so that we avoid a premature merge. If after reading the feedback you really think it’s ready to go, please feel free to re-apply the labels.

**Comment by mpickering** [comment]
It seems that the validate script is built with the same version of GHC as the one which is being tested.

It would be better in my opinion if the validate script was decoupled from the version of GHC being tested. (This was achieved when it was a bash script)

It would be bad to get into a situation where adding testing for a new compiler is blocked because one of the dependencies of the validate script hasn't been updated yet.

**Comment by chreekat** [comment]
> It seems that the validate script is built with the same version of GHC as the one which is being tested.

On that note, the script depends on `terminal-size`, which depends on `ghc-prim`, which "is an internal package, only for the use of GHC developers. GHC users should not use it!" Perhaps that dependency could be dropped, at the very least?

**Comment by geekosaur** [comment]
Is that not available from `ansi-terminal`? If not, why does `terminal-size` depend explicitly on `ghc-prim`? The correct way to get terminal size doesn't require GHC internals, afaik. (Granting that the way I know is POSIX; I don't know what Windows requires, but I also am not sure you get any options there, I recall it resizing fonts to keep the 80x25 size.)

**Comment by 9999years** [comment]
> On that note, the script depends on terminal-size, which depends on ghc-prim, which "is an internal package, only for the use of GHC developers. GHC users should not use it!" Perhaps that dependency could be dropped, at the very least?

`terminal-size` only depends on `ghc-prim` when built with GHC between 7.2 and 7.6, which I suspect we do not have to worry about:

https://github.com/biegunka/terminal-size/blob/8c5765027ae6194191d6da11d69a49c37a78a73a/terminal-size.cabal#L32-L34

**Comment by 9999years** [comment]
> It seems that the validate script is built with the same version of GHC as the one which is being tested.
>
> It would be better in my opinion if the validate script was decoupled from the version of GHC being tested. (This was achieved when it was a bash script)

Nothing requires this to be the case; we could easily do `cabal build --with-ghc ... cabal-validate` in CI or locally. The 3-line `validate.sh` entrypoint is simply provided as a convenience.

**Comment by 9999years** [comment]
> It would be bad to get into a situation where adding testing for a new compiler is blocked because one of the dependencies of the validate script hasn't been updated yet.

The new dependencies are `ansi-terminal`, `terminal-size`, and `typed-process`. All of the others are either bundled with GHC or used by other components in this repository already.

`ansi-terminal` could be removed pretty easily; we only use it to write ANSI escape sequences.

`terminal-size` only depends on `base >=4 && <5` (except on Windows, where it depends on `process` and `Win32`, both bundled with GHC), and it's only 120 lines of code anyways, so I think that's fine.

`typed-process` pulls in `unliftio-core` (all other dependencies are bundled with GHC or included in other Cabal components already), which also seems pretty reasonable. We _could_ remove this, but the ergonomics and maintainability of `cabal-validate` would suffer (IMO) and we'd have to devote a fair amount of effort to reproducing its error messages.

File: cabal-validate/src/Main.hs
Comment (9999years) [reply cabal-validate/src/Main.hs:127]: Yes, it's definitely not one of Haskell's strong suits. I originally tried to use `OverloadedRecordDot`, but of course we run CI with versions of GHC that don't support it.

I also want this script to be [as accessible as possible](https://www.haskellforall.com/2015/09/how-to-make-your-haskell-code-more.html), and `NamedFieldPuns` make it less clear where variables are declared.

I do agree the verbosity is unideal, though.

File: cabal-validate/src/Cli.hs
Comment (9999years) [reply cabal-validate/src/Cli.hs:62]: Me too, but `fourmolu` removes it automatically!

File: cabal-validate/src/Main.hs
Comment (9999years) [reply cabal-validate/src/Main.hs:364]: I think that would reduce readability for negligible gain.

File: cabal-validate/src/Main.hs
Comment (9999years) [reply cabal-validate/src/Main.hs:207]: > The doctest step in validate.sh (which is what you were inspired by, I guess) has bit rotted.

To be clear: Nothing in this PR is 'inspired', this is a 1:1 translation of `validate.sh`, including many silly details. I'd love to work on fixing it up once this PR is merged, but for now I want to keep the `validate.sh` behavior unchanged.

**Comment by 9999years** [comment]
> My only big concern is how many times it will have to be built from scratch and whether it will bring a noticeable slowdown in CI.

It seems like [the `print-config` step](https://github.com/haskell/cabal/actions/runs/10780535402/job/29896512707?pr=10320), which builds `cabal-validate`, finishes in about 17 seconds in CI. (Judging by local runs, about ~1 second or less of that is running the `cabal-validate` script.) That seems like a reasonable cost to me, and from the fact that none of the dependencies are built in that step, it seems like the dependencies are cached as well, so we can expect to not pay that cost unless we're changing `validate.sh`.

**Comment by chreekat** [comment]
> `terminal-size` only depends on `ghc-prim` when built with GHC between 7.2 and 7.6, which I suspect we do not have to worry about:

Oh, thanks for pointing that out! I only got as far as the package home page on Hackage, which lists the dependencies without mentioning the constraints.

**Comment by 9999years** [comment]
@ulysses4ever @mpickering I'd appreciate if you could take a look at my responses and see if it looks good or if there's anything else we should address before merge. Thanks!

**Comment by ulysses4ever** [comment]
Yes, sorry, I meant to post it yesterday: I don't have any other comments and think it can be handed over to the merge bot.

**Comment by geekosaur** [comment]
Switched the merge label: mergify just balked at queueing it.

**Comment by geekosaur** [comment]
@mergify backport 3.12

**Pull request merged**
